app/lib/features/chat_ng/providers/notifiers/chat_room_messages_notifier.dart

```diff
@@ -0,0 +1,202 @@
+import 'dart:async';
+
+import 'package:acter/common/extensions/options.dart';
+import 'package:acter/features/chat_ng/models/chat_room_state/chat_room_state.dart';
+import 'package:acter/features/chat/providers/chat_providers.dart';
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:flutter/widgets.dart';
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod/riverpod.dart';
+
+final _log = Logger('a3::chat::room_notifier');
+
+class ChatRoomMessagesNotifier extends StateNotifier<ChatRoomState> {
+  final Ref ref;
+  final String roomId;
+  late TimelineStream timeline;
+  late Stream<RoomMessageDiff> _listener;
+  late StreamSubscription<RoomMessageDiff> _poller;
+
+  ChatRoomMessagesNotifier({
+    required this.roomId,
+    required this.ref,
+  }) : super(const ChatRoomState()) {
+    _init();
+  }
+
+  Future<void> _init() async {
+    try {
+      timeline = await ref.read(timelineStreamProvider(roomId).future);
+      _listener = timeline.messagesStream(); // keep it resident in memory
+      _poller = _listener.listen(
+        (diff) {
+          state = handleDiff(state, diff);
+        },
+        onError: (e, s) {
+          _log.severe('msg stream errored', e, s);
+        },
+        onDone: () {
+          _log.info('msg stream ended');
+        },
+      );
+      ref.onDispose(() => _poller.cancel());
+      do {
+        await loadMore(failOnError: true);
+        await Future.delayed(const Duration(milliseconds: 200), () => null);
+      } while (state.hasMore && state.messageList.length < 10);
+    } catch (e, s) {
+      _log.severe('Error loading more messages', e, s);
+      state = state.copyWith(
+        loading: ChatRoomLoadingState.error(e.toString()),
+      );
+    }
+  }
+
+  Future<void> loadMore({bool failOnError = false}) async {
+    if (state.hasMore && !state.loading.isLoading) {
+      try {
+        state = state.copyWith(
+          loading: const ChatRoomLoadingState.loading(),
+        );
+        final hasMore = !await timeline.paginateBackwards(20);
+        // wait for diffRx to be finished
+        state = state.copyWith(
+          hasMore: hasMore,
+          loading: const ChatRoomLoadingState.loaded(),
+        );
+      } catch (e, s) {
+        _log.severe('Error loading more messages', e, s);
+        state = state.copyWith(
+          loading: ChatRoomLoadingState.error(e.toString()),
+        );
+        if (failOnError) {
+          rethrow;
+        }
+      }
+    }
+  }
+}
+
+@visibleForTesting
+ChatRoomState handleDiff(ChatRoomState state, RoomMessageDiff diff) {
+  final action = diff.action();
+  switch (action) {
```

Comment:
check the tests for more. but the key thing is : we have both a list of ids and the messages in a hashmap now.